lib/puppet/type/deferred_resources.rb

```diff
@@ -72,7 +79,68 @@
 
     validate do |value|
       unless  value.is_a?(Hash) || value.is_a?(Array)
-        raise 'Expecting a Hash or Array for :resources'
+        raise Puppet::Error, 'Expecting a Hash or Array for :resources'
+      end
+    end
+  end
+
+  newparam(:override_existing_attributes) do
+    desc <<-EOM
+     A Hash or Array of items that should be updated on existing attributes if
+     they exist.
+
+      This is basically a controlled resource collector and absolutely must not
+      be taken lightly when used since it will affect existing resources in
+      your catalog.
+
+      If you want to be explicit, use a Resource Collector and do not set this.
+
+      If a Hash is passed, each key is the attribute that can be overridden and
+      an optional hash of options can be passed with the following meanings.
+
+        * 'invalidates':
+          * An Array of entries that this particular parameter invalidates.
+            This means that the items in the list will be set to `nil` in the
+            overridden resource.
+    EOM
+
+    munge do |value|
+      if value.is_a?(Array)
+        value = Hash[value.map{|x| [x, {}]}]
+      end
+
+      value.keys.each do |k|
+        value[k] = {} if value[k].nil?
+      end
+
+      value
+    end
+
+    validate do |value|
+      unless (value.is_a?(Hash) || value.is_a?(Array)) && !value.empty?
+        raise Puppet::Error, 'Expecting an Array or Hash with contents for :override_existing_attributes'
+      end
+
+      valid_control_opts = [
+        'invalidates'
+      ]
+
+      if value.is_a?(Hash)
+        value.each_pair do |attr, opts|
+          if attr
+            if attr.is_a?(Hash)
```

Comment:
Should you be checking the key (attr) here or the value (opts).   I went in and edited the hash in the code to contain what I thought was invalid values and it passed.

Reply:
@jeannegreulich Great catch! Working on getting a test in for that.